python/dgl/nn/pytorch/glob.py

```diff
@@ -0,0 +1,673 @@
+        graph.ndata.pop(value_name)
+        graph.ndata.pop(out_name)
+        graph.edata.pop(att_name)
+
+        # Position-wise Feed Forward Network
+        feat = self.norm_inter(feat + self.ffn(feat))
+
+        return feat
+
+
+class SetAttentionBlock(nn.Module):
+    r""" SAB block mentioned in Set-Transformer paper."""
+    def __init__(self, d_model, num_heads, d_head, d_ff, dropouth=0., dropouta=0.):
+        super(SetAttentionBlock, self).__init__()
+        self.mha = MultiHeadAttention(d_model, num_heads, d_head, d_ff,
+                                      dropouth=dropouth, dropouta=dropouta)
+
+    def forward(self, feat, graph, sab_graph=None):
+        """
+        Compute a Set Attention Block.
+
+        Parameters
+        ----------
+        feat : torch.Tensor
+            The input feature
+        graph : DGLGraph
+            The graph.
+        sab_graph: DGLGraph or None
+            The graph to apply Message Passing on, set to None if not
+            specified.
+        """
+        if sab_graph is None:
+            sab_graph = _create_fully_connected_graph(graph)
+
+        q_nids = th.arange(sab_graph.number_of_nodes())
+        kv_nids = q_nids
+
+        return self.mha(sab_graph, feat, feat, q_nids, kv_nids)
+
+class InducedSetAttentionBlock(nn.Module):
+    r""" ISAB block mentioned in Set-Transformer paper."""
+    def __init__(self, m, d_model, num_heads, d_head, d_ff, dropouth=0., dropouta=0.):
+        super(InducedSetAttentionBlock, self).__init__()
+        self.m = m
+        self.I = nn.Parameter(
+            th.FloatTensor(m, d_model)
+        )
+        self.mha = nn.ModuleList([
+            MultiHeadAttention(d_model, num_heads, d_head, d_ff,
+                               dropouth=dropouth, dropouta=dropouta) for _ in range(2)])
+        self._reset_parameters()
+
+    def _reset_parameters(self):
+        nn.init.xavier_uniform_(self.I)
+
+    def forward(self, feat, graph, isab_graph=None):
+        """
+        Compute an Induced Set Attention Block.
+
+        Parameters
+        ----------
+        feat : torch.Tensor
+            The input feature
+        graph : DGLGraph
+            The graph.
+        isab_graph: DGLGraph or None
+            The graph to apply Message Passing on, set to None if not
+            specified.
+        """
+        query = self.I
+        batch_size = 1
+        if isinstance(graph, BatchedDGLGraph):
+            batch_size = graph.batch_size
+
+        if isab_graph is None:
+            induced_graph = _create_graph_from_num_nodes([self.m] * batch_size)
+            isab_graph = [
+                _create_bipartite_graph(graph, induced_graph),
+                _create_bipartite_graph(induced_graph, graph)
+            ]
+
+        query = query.repeat(batch_size, 1)
+        for mha, (g, kv_nids, q_nids) in zip(self.mha, isab_graph):
+            rst = mha(g, query, feat, q_nids, kv_nids)
+            query, feat = feat, rst
```

Comment:
I would not use loop here if there are only two elements. This makes it hard to understand the codes. The math is quite clear, so try follow the equation (just an example):
```
H = self.mab1(I, X)
Z = self.mab2(X, H)
```

Reply:
DONE